File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        return twoPointer(nums);
        return bruteforce(nums);
    }
private:
    // O(n) Time and O(n) extra space
    vector<int> twoPointer(vector<int>& nums){
        int i = 0, j = nums.size() - 1;
        vector<int> res;
        
        while(i <= j){
            if(abs(nums[i]) < abs(nums[j])){
                res.push_back(nums[j] * nums[j]);
                j--;
            }
            else{
                res.push_back(nums[i] * nums[i]);
                i++;
            }
            
        }
        reverse(res.begin(), res.end());
        return res;
        
    }
    
    // O(nlogn) Time and O(1) extra space
    vector<int> bruteforce(vector<int>& nums){
        for(int i = 0; i < nums.size(); i++){
            nums[i] = nums[i] * nums[i];
        }
        sort(nums.begin(), nums.end());
        return nums;
    }
};

File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.py
class Solution:
    # O(nlogn) Time and O(1) Extra space
    def bruteForce(self, nums):
        for i in range(len(nums)):
            nums[i] = nums[i] * nums[i]
        nums.sort()
        return nums    
    
    # O(n) Time and O(n) extra Space
    def optimised(self, nums):
        result = []
        
        i, j = 0, len(nums) - 1
        
        while i <= j:
            if abs(nums[i]) > abs(nums[j]):
                result.append(nums[i] * nums[i])
                i += 1
            else:
                result.append(nums[j] * nums[j])
                j -= 1
        
        return reversed(result)
    
    def sortedSquares(self, nums: List[int]) -> List[int]:
        # return self.bruteForce(nums)    
        return self.optimised(nums)